actions: provision python from conda-forge

oliver-sanders opened the pull request:
* Provision Python , Bash and other deps from conda-forge.
* Divorce from the OS environment and give us more flexibility in the software versions we utilise.
* Reduce usage of `apt get` (to `at`) and remove usage `brew install` completely.
* Fix SQL statements in the functional tests to comply with changes in sqlite3 3.46.0+ (closes https://github.com/cylc/cylc-flow/issues/6630)
* Fix a functional test (wasn't failing before for unknown reasons, but the fix makes sense)
* Install subversion to ensure `svn` tests aren't skipped.
* Fix installation of the mock `mail` command to ensure mail tests aren't skipped.

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] Changelog entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

MetRonnie commented on .github/workflows/test_functional.yml:
```diff
@@ -56,34 +60,34 @@ jobs:
           # macos
           - name: 'macos 1/5'
             os: 'macos-latest'
-            python-version: '3.9'
+            python-version: '3.8'  # oldest available
             test-base: 'tests/f'
             chunk: '1/5'
             platform: '_local_background*'
           - name: 'macos 2/5'
             os: 'macos-latest'
-            python-version: '3.9'
+            python-version: '3'  # newest available
```
The original intention was to run the `1/4` chunk in MacOS on the oldest Python version. (However at some point this was split into `1/5` and `2/5` for faster completion.)

What is the intention behind running different chunks in different Python versions?

oliver-sanders replied:
I thought we should probably have a test for latest Python on Mac OS as a canary. Happy to revert (unrelated really).

oliver-sanders commented on .github/workflows/test_functional.yml:
```diff
@@ -167,6 +157,8 @@ jobs:
       - name: Configure Atrun
         if: contains(matrix.platform, '_local_at')
         run: |
+          sudo apt-get update
+          sudo apt-get install -y at
```
`at` is the one thing we need to get from the OS package manager.

oliver-sanders commented on .github/workflows/test_fast.yml:
```diff
@@ -12,6 +12,10 @@ concurrency:
 
+defaults:
+  run:
+    shell: bash -c "exec $CONDA_PREFIX/bin/bash -elo pipefail {0}"
```
Setting `-eo pipefail` for error checking.

MetRonnie commented on .github/workflows/test_fast.yml:
```diff
@@ -39,24 +41,26 @@ jobs:
+            python=${{ matrix.python-version }}
+            pip
+            bash
+            sqlite
```
I've noticed the SVN unit tests are being skipped. Which is strange because we didn't install SVN before. Is it possible to include SVN here?

oliver-sanders replied:
Good spotting.

This suggests that if subversion is installed, the `svn` executable is no longer in the `$PATH`.

I'm not sure if this is a consequence of the switch to micromamba (doesn't make sense) or a consequence of dropping `apt-get` installation (doesn't make sense).

oliver-sanders replied:
We could just do this:

```suggestion
            sqlite
            subversion
```

oliver-sanders replied:
> Ah wait, it's because it was installed on Ubuntu 22 but not latest. Is it possible to include SVN here?

I have just discovered this myself.

Let's try installing it from conda-forge on the newer images.

oliver-sanders replied:
Done, will see if it works...

MetRonnie commented:
I can't remember where you got to with the mail command, but looks like it's not working
```
[18:37:31] tests/f/events/13-task-event-mail-globalcfg.t .......... skipped: "mail" command not available
```

oliver-sanders commented:
> Looks like another sqlite quote issue (although test passing for some reason)

Fixed and turned the `sqlite3` command calls into `run_ok` tests to make sure they can't fail silently.

oliver-sanders commented:
> I can't remember where you got to with the mail command, but looks like it's not working:

The mock mail implementation is added to the `$PATH` in the `.bashrc` file. I've got the action to `cat` the `.bashrc` file so we can see what's going on here.

oliver-sanders commented:
Ok, comments addressed:

* Fixed the mocked `mail` command (I think this was broken before). The mail tests are now running [e.g.](https://github.com/cylc/cylc-flow/actions/runs/13787511565/job/38561716670?pr=6655#step:14:226).
* Installed the `svn` command (this was broken by GH image changes).
* Fixed an SQL statement and updated the test to ensure the failure of this operation fails the test.

MetRonnie commented on .github/workflows/test_functional.yml:
```diff
@@ -128,36 +144,12 @@ jobs:
       - name: Add .github/bin/ to PATH
         # Sets up mocked mail command & any other custom executables
-        run: echo "${{ github.workspace }}/.github/bin" >> $GITHUB_PATH
+        # run: echo "${{ github.workspace }}/.github/bin" >> $GITHUB_PATH
```
```suggestion
        # Adding to $GITHUB_PATH does not work when using setup-micromamba and/or login shell
```

MetRonnie reviewed (approved):
Nice. Got one comment suggestion that I would rather have merged but not a blocker

oliver-sanders commented:
Done.
